toolchain/semantics/semantics_handle_index.cpp

```diff
@@ -5,17 +5,56 @@
 #include <utility>
 
 #include "toolchain/semantics/semantics_context.h"
+#include "toolchain/semantics/semantics_node.h"
+#include "toolchain/semantics/semantics_node_kind.h"
 
 namespace Carbon {
 
 auto SemanticsHandleIndexExpressionStart(SemanticsContext& context,
                                          ParseTree::Node parse_node) -> bool {
-  return context.TODO(parse_node, "HandleIndexExpressionStart");
+  context.node_stack().Push(parse_node, context.node_stack().PopExpression());
+  return true;
 }
 
 auto SemanticsHandleIndexExpression(SemanticsContext& context,
                                     ParseTree::Node parse_node) -> bool {
-  return context.TODO(parse_node, "HandleIndexExpression");
+  auto ind_node = context.node_stack().PopExpression();
+  auto [ind_expr_parse_node, name_node_id] =
+      context.node_stack()
+          .PopWithParseNode<ParseNodeKind::IndexExpressionStart>();
+  auto name_node = context.semantics_ir().GetNode(name_node_id);
+
+  if (context.semantics_ir().GetTypeAllowBuiltinTypes(name_node.type_id()) !=
+      SemanticsNodeId::BuiltinError) {
```

Comment:
Why do you call GetType for this, instead of comparing with SemanticsTypeId::Error directly?